src/lib/support/IntrusiveList.h

```diff
@@ -0,0 +1,222 @@
+        IteratorBase(IteratorBase &&)      = default;
+        IteratorBase & operator=(const IteratorBase &) = default;
+        IteratorBase & operator=(IteratorBase &&) = default;
+
+        bool operator==(const IteratorBase & that) const { return mCurrent == that.mCurrent; }
+        bool operator!=(const IteratorBase & that) const { return !(*this == that); }
+
+        IteratorBase & operator++()
+        {
+            mCurrent = mCurrent->mNext;
+            return *this;
+        }
+
+        IteratorBase operator++(int)
+        {
+            IteratorBase res(mCurrent);
+            mCurrent = mCurrent->mNext;
+            return res;
+        }
+
+        IteratorBase & operator--()
+        {
+            mCurrent = mCurrent->mPrev;
+            return *this;
+        }
+
+        IteratorBase operator--(int)
+        {
+            IteratorBase res(mCurrent);
+            mCurrent = mCurrent->mPrev;
+            return res;
+        }
+
+    protected:
+        IntrusiveListNodeBase * mCurrent;
+    };
+
+    bool Empty() const { return mNode.mNext == &mNode; }
+
+    void Erase(IteratorBase pos) { pos.mCurrent->Remove(); }
+
+protected:
+    // The list is formed as a ring with mNode being the end.
+    //
+    // begin                   end
+    //   v                      v
+    // item -> item -> ... -> mNode
+    //   ^                      |
+    //    \--------------------/
+    //
+    IntrusiveListBase() : mNode(&mNode, &mNode) {}
+    ~IntrusiveListBase() { mNode.Remove(); }
+
+    IteratorBase begin() { return IteratorBase(mNode.mNext); }
+    IteratorBase end() { return IteratorBase(&mNode); }
+
+    void PushFront(IntrusiveListNodeBase * node) { mNode.Append(node); }
+    void PushBack(IntrusiveListNodeBase * node) { mNode.Prepend(node); }
+
+    void InsertBefore(IteratorBase pos, IntrusiveListNodeBase * node)
+    {
+        VerifyOrDie(pos.mCurrent->IsInList());
+        pos.mCurrent->Prepend(node);
+    }
+
+    void InsertAfter(IteratorBase pos, IntrusiveListNodeBase * node)
+    {
+        VerifyOrDie(pos.mCurrent->IsInList());
+        pos.mCurrent->Append(node);
+    }
+
+    void Remove(IntrusiveListNodeBase * node)
+    {
+        VerifyOrDie(Contains(node));
+        node->Remove();
+    }
+
+    bool Contains(IntrusiveListNodeBase * node)
+    {
+        for (auto & iter : *this)
+        {
+            if (&iter == node)
+                return true;
+        }
+        return false;
+    }
+
+private:
+    IntrusiveListNodeBase mNode;
+};
+
+template <typename T>
+class IntrusiveListBaseHook
+{
+public:
+    static_assert(std::is_base_of<IntrusiveListNodeBase, T>::value, "T must be derived from IntrusiveListNodeBase");
+    static T * ToObject(IntrusiveListNodeBase * node) { return static_cast<T *>(node); }
+    static IntrusiveListNodeBase * ToNode(T * object) { return static_cast<IntrusiveListNodeBase *>(object); }
+};
+
+template <typename T, typename Hook = IntrusiveListBaseHook<T>>
```

Comment:
some doccomments about what this does and how it is used would be useful.

Reply:
From the name it already says that it is intrusive and type T is expected to be the data type. Consider something like this:

```
/// A double-linked list where the data is stored together with the previous/next pointers for cache efficiency
/// and compactness.
///
/// The node  type (T) MUST inherit from IntrusiveListNodeBase (this contains the previous/next 
/// pointers and operations around them.
///
/// Node lifetime is assumed to be longer than the list they belong to. A list is effcively a single node into
/// a chain of other nodes referenced by pointers.
///
/// A node may only belong to a single list. The code will assert (via VerifyOrDie) on this invariant.
///
/// Example usage:
/// 
///     class ListNode : public IntrusiveListNodeBase  {};
///     // ...
///     ListNode a,b,c;
///     IntrusiveList<ListNode> list;    // NOTE: node lifetime >= list lifetime
/// 
///     list.PushBack(&a);
///     list.PushFront(&b);
///     assert(list.Contains(&a)  && list.Contains(&b) && !list.Contains(&c));
```

which tells you more info:
   - that we inherit from a base to have previous/next pointers
   - that code will VerifyOrDie on belonging to a single list
   - that the list is double-linked
   
without the need to read the code.